Add support for health_overview API endpoint to Tractive integration

BenWolstencroft opened the pull request:
## Breaking change

N/A - This change maintains backward compatibility during Tractive's API transition period.

## Proposed change

This PR updates the Tractive integration to support the new `health_overview` API endpoint, which replaces the deprecated `wellness_overview` endpoint. The changes ensure all health and activity sensors continue to work correctly with the new API.

Key changes:
- Fetch `health_overview` data on integration startup
- Handle `health_overview` events from Tractive's event stream
- Dispatch initial health data to sensors after platform setup to prevent "unavailable" state on startup
- Maintain backward compatibility with `wellness_overview` events during the transition period
- Add `trackable_object()` method to `TractiveClient` for accessing pet-specific data

The implementation has been tested on Home Assistant OS with a real Tractive device and all sensors populate correctly.

## Type of change

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] Dependency upgrade
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #151651
- This PR is related to issue: zhulik/aiotractive#39
- Link to documentation pull request: N/A (no user-facing changes)
- Link to developer documentation pull request: N/A
- Link to frontend pull request: N/A
- Link to aiotractive library PR: https://github.com/zhulik/aiotractive/pull/49

**Note:** This PR depends on https://github.com/zhulik/aiotractive/pull/49 being merged first.

## Checklist

- [x] I understand the code I am submitting and can explain how it works.
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.
- [x] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

BenWolstencroft commented:
Theres pending code changes for this pr to manifests once the version number for updated aiotractive is available, unable to merge yet.

bieniu commented on homeassistant/components/tractive/__init__.py:
```diff
@@ -226,9 +251,13 @@ async def _listen(self) -> None:
                     if server_was_unavailable:
                         _LOGGER.debug("Tractive is back online")
                         server_was_unavailable = False
+                    # Support both deprecated wellness_overview and new health_overview
```
Tractive has deprecated the `wellness_overview` message (the API no longer sends these messages), we should remove all code related to `wellness_overview`.

BenWolstencroft replied:
Agreed, i didn't want to 'overstep' :D

BenWolstencroft commented on homeassistant/components/tractive/sensor.py:
```diff
@@ -175,6 +181,39 @@ def handle_status_update(self, event: dict[str, Any]) -> None:
+        key=ATTR_HEALTH_ALERTS,
+        translation_key="health_alerts",
+        signal_prefix=TRACKER_WELLNESS_STATUS_UPDATED,
+        state_class=SensorStateClass.MEASUREMENT,
```
You're absolutely right, i'll split the PR

bieniu commented on homeassistant/components/tractive/__init__.py:
```diff
@@ -296,24 +309,34 @@ def _send_switch_update(self, event: dict[str, Any]) -> None:
         }
         self._dispatch_tracker_event(
-            TRACKER_WELLNESS_STATUS_UPDATED, event["pet_id"], payload
+            TRACKER_WELLNESS_STATUS_UPDATED, data.get("petId"), payload
```
```suggestion
            TRACKER_WELLNESS_STATUS_UPDATED, data["petId"], payload
```
We should not use `.get()` here, ID is a string.

bieniu commented on homeassistant/components/tractive/__init__.py:
```diff
@@ -296,24 +309,34 @@ def _send_switch_update(self, event: dict[str, Any]) -> None:
+            sleep_day = sleep.get("minutesDaySleep")
+            sleep_night = sleep.get("minutesNightSleep")
+            # Calm minutes can be used as rest indicator
+            minutes_rest = sleep.get("minutesCalm")
```
```suggestion
        sleep_day = sleep.get("minutesDaySleep")
        sleep_night = sleep.get("minutesNightSleep")
        # Calm minutes can be used as rest indicator
        minutes_rest = sleep.get("minutesCalm")
```

bieniu commented on homeassistant/components/tractive/sensor.py:
```diff
@@ -125,13 +121,6 @@ def handle_status_update(self, event: dict[str, Any]) -> None:
         signal_prefix=TRACKER_WELLNESS_STATUS_UPDATED,
         state_class=SensorStateClass.TOTAL,
     ),
-    TractiveSensorEntityDescription(
```
Removing deprecated entities should not be part of this PR.

BenWolstencroft committed:
Add support for health_overview API endpoint to Tractive integration

- Fetch health_overview data on startup (replaces deprecated wellness_overview)
- Add support for health_overview events from Tractive API
- Dispatch initial health data to sensors after platform setup
- Add trackable_object() method to TractiveClient for accessing pet data
- Remove deprecated sensors no longer provided by API:
  - activity_label (activity quality indicator)
  - calories (calories burned)
  - sleep_label (sleep quality indicator)
- Update tests with health_overview mocks and fixtures

Related to zhulik/aiotractive#39

BenWolstencroft committed:
Apply suggestions from code review

Co-authored-by: Maciej Bieniek <bieniu@users.noreply.github.com>

BenWolstencroft committed:
Restore deprecated sensors (activity_label, calories, sleep_label)

BenWolstencroft committed:
Restore _send_wellness_update for backward compatibility with wellness_overview events

BenWolstencroft committed:
Remove wellness-only fields from health_overview payload

bieniu commented on homeassistant/components/tractive/__init__.py:
```diff
@@ -316,6 +337,27 @@ def _send_wellness_update(self, event: dict[str, Any]) -> None:
+            ATTR_MINUTES_REST: sleep.get("minutesCalm"),
+        }
+        self._dispatch_tracker_event(
+            TRACKER_WELLNESS_STATUS_UPDATED, data["petId"], payload
```
We should use different prefix for this event. And sensors `minutes_active`, `minutes_rest`, `daily_goal`, `minutes_day_sleep` and `minutes_night_sleep` should use this prefix. Otherwise we will get a `KeyError`.
```suggestion
            TRACKER_HEALTH_OVERVIEW_STATUS_UPDATED, data["petId"], payload
```

BenWolstencroft replied:
fixed

BenWolstencroft committed:
Add TRACKER_HEALTH_OVERVIEW_UPDATED constant and rename trackable_object methods

bieniu commented on homeassistant/components/tractive/__init__.py:
```diff
@@ -144,17 +151,23 @@ async def _generate_trackables(
+        tracker.details(),
+        tracker.hw_info(),
+        tracker.pos_report(),
+        trackable_obj.health_overview(),
```
```suggestion
        trackable_pet.health_overview(),
```

bieniu commented on homeassistant/components/tractive/const.py:
```diff
@@ -28,6 +28,7 @@
 TRACKER_POSITION_UPDATED = f"{DOMAIN}_tracker_position_updated"
 TRACKER_SWITCH_STATUS_UPDATED = f"{DOMAIN}_tracker_switch_updated"
 TRACKER_WELLNESS_STATUS_UPDATED = f"{DOMAIN}_tracker_wellness_updated"
+TRACKER_HEALTH_OVERVIEW_UPDATED = f"{DOMAIN}_tracker_health_overview_updated"
```
Sensors `minutes_active`, `minutes_rest`, `daily_goal`, `minutes_day_sleep` and `minutes_night_sleep` need to use this prefix, e.g.:

```python
    TractiveSensorEntityDescription(
        key=ATTR_MINUTES_NIGHT_SLEEP,
        translation_key="minutes_night_sleep",
        native_unit_of_measurement=UnitOfTime.MINUTES,
        signal_prefix=TRACKER_HEALTH_OVERVIEW_UPDATED,
        state_class=SensorStateClass.TOTAL,
    ),
```

BenWolstencroft commented on homeassistant/components/tractive/__init__.py:
```diff
@@ -144,17 +151,23 @@ async def _generate_trackables(
         return None
 
     tracker = client.tracker(trackable["device_id"])
+    trackable_obj = client.trackable_object(trackable["_id"])
```
`client` in this instance is an instance of `aiotractive.Tractive`, **not** an instance of the `TractiveClient` within this codebase, the `TractiveClient.trackable_pet` (was `TractiveClient.trackable_object` pre my change) appears to be completely unused and should be removed in a future PR.

I suggest that separately some improvements to the aiotractive library can be made first where we sunset unclear names whilst maintaining backwards compatibility for a sunset period before removal - to allow for cleaner usage in clients.

However for clarity I will rename the trackable_obj variable

BenWolstencroft committed:
Use health_overview signal for activity and sleep sensors

BenWolstencroft committed:
improve variable name

BenWolstencroft committed:
revert definition rename

BenWolstencroft committed:
Remove unused trackable_pet method from TractiveClient class

bieniu reviewed (approved):
Tested with real devices, works as intended. Great work @BenWolstencroft👍
